.github/actions/setup-env/action.yml

```diff
@@ -0,0 +1,17 @@
+name: Setup Env
+description: Setup Node.js and install dependencies
+
+runs:
+  using: composite
+  steps:
+    - name: Setup Node.js
+      id: setup-node
+      uses: actions/setup-node@v4
+      with:
+        node-version-file: .node-version
+        cache: npm
+
+    - name: Install Dependencies
+      id: install
+      shell: bash
+      run: npm ci
```

Comment:
Happy to explain! That's correct, `esbuild` will bundle all of an app's dependencies into a single JS file. It also tree-shakes dependencies so if there's any code that you don't use, that will not make it into the final bundle too.

However all of our CI checks operate on the source code, so dependencies need to be installed so that the checks run correctly. For example, ESLint will also perform basic import validation and type checking, which requires that dependencies installed. Also, since we're running tests directly using the source code, dependencies must be installed so that they can resolve correctly inside of the test runtime. This is unique to the way JS operates so I can imagine it'd be different to Go.

Hope that helps explain things!

Reply:
Thanks for the background! According to https://github.com/evanw/esbuild/issues/2713 ESBuild includes any dependency that is used, independent of whether it's listed as `dependency` or `devDependency` in `package.json`. So, if we keep it clean, the JS bundle of this action does not contain the devDependencies since they aren't used by our "production" code. TIL